scio-grpc/src/main/java/com/spotify/scio/grpc/GrpcDoFn.java

```diff
@@ -0,0 +1,127 @@
+/*
+ * Copyright 2022 Spotify AB
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.spotify.scio.grpc;
+
+import com.google.common.util.concurrent.ListenableFuture;
+import com.spotify.scio.transforms.GuavaAsyncLookupDoFn;
+import io.grpc.Channel;
+import io.grpc.stub.AbstractStub;
+import org.apache.beam.sdk.transforms.SerializableBiFunction;
+import org.apache.beam.sdk.transforms.SerializableFunction;
+
+import java.io.Serializable;
+import java.util.function.Supplier;
+
+import static java.util.Objects.requireNonNull;
+
+/**
+ * DoFn that makes API calls over a managed GRPC channel.
+ *
+ * @param <RequestT>
+ * @param <ResponseT>
+ * @param <ClientT>
+ */
+public class GrpcDoFn<RequestT, ResponseT, ClientT extends AbstractStub<ClientT>>
+    extends GuavaAsyncLookupDoFn<RequestT, ResponseT, ClientT> {
+  static int DEFAULT_MAX_PENDING_REQUESTS = 1000;
+
+  private final ChannelSupplier channelSupplier;
+  private final SerializableFunction<Channel, ClientT> newClientFn;
+  private final SerializableBiFunction<ClientT, RequestT, ListenableFuture<ResponseT>> lookupFn;
+
+  GrpcDoFn(
+      ChannelSupplier channelSupplier,
+      SerializableFunction<Channel, ClientT> newClientFn,
+      SerializableBiFunction<ClientT, RequestT, ListenableFuture<ResponseT>> lookupFn,
+      Integer maxPendingRequests) {
+    super(maxPendingRequests, new NoOpCacheSupplier<>());
+    this.channelSupplier = channelSupplier;
+    this.newClientFn = newClientFn;
+    this.lookupFn = lookupFn;
+  }
+
+  @Override
+  public ResourceType getResourceType() {
+    // gRPC stubs are thread safe, we can share the client per instance
+    return ResourceType.PER_INSTANCE;
+  }
```

Comment:
This should be a nice optimization. I feel we can create the interceptors directly in the `newClientFn` if need be since the client will be shared anyway